### ISSUE TITLE
[BUG] DWA Compression for single channel EXR images only works with tiles

### ISSUE THREAD

**Issue by RoseHirigoyen** [OpenImageIO/oiio#3821]
**Describe the bug**
For single channel EXR images, I understand that DWAA/DWAB compression only works for images bigger than 16 x 16 and if the width and height are powers of two (from [this line of code](https://github.com/OpenImageIO/oiio/blob/571a16717f250ff0d11dad6e72c431fcf39b05f1/src/openexr.imageio/exroutput.cpp#L740)). However, in the way it is implemented, it only works for tiled images.

**To Reproduce**
The compression level of this image is zip (by default because it is not tiled)
```cmd
% oiiotool --pattern checker 1024x512 1 -d half -compression dwaa -o test.exr
% oiiotool --info -v -a test.exr
Reading test.exr
test.exr             : 1024 x  512, 1 channel, half openexr
    channel list: Y
    compression: "zip"
    DateTime: "2023:04:26 12:25:09"
    PixelAspectRatio: 1
    screenWindowCenter: 0, 0
    screenWindowWidth: 1
    Software: "OpenImageIO 2.4.9.0 : oiiotool --pattern checker 1024x512 1 -d half -compression dwaa -o test.exr"
    Exif:ImageHistory: "oiiotool --pattern checker 1024x512 1 -d half -compression dwaa -o test.exr"
    oiio:ColorSpace: "Linear"
    oiio:subimages: 1
```

The compression level of this image is dwaa
```
% oiiotool --pattern checker 1024x512 1 -d half -compression dwaa -tile 1024 512 -o test.exr
# compression is dwaa
% oiiotool --info -v -a test.exr
Reading test.exr
test.exr             : 1024 x  512, 1 channel, half openexr
    channel list: Y
    tile size: 1024 x 512
    compression: "dwaa"
    DateTime: "2023:04:26 12:25:31"
    PixelAspectRatio: 1
    screenWindowCenter: 0, 0
    screenWindowWidth: 1
    Software: "OpenImageIO 2.4.9.0 : oiiotool --pattern checker 1024x512 1 -d half -compression dwaa -tile 1024 512 -o test.exr"
    Exif:ImageHistory: "oiiotool --pattern checker 1024x512 1 -d half -compression dwaa -tile 1024 512 -o test.exr"
    oiio:ColorSpace: "Linear"
    oiio:subimages: 1
    openexr:dwaCompressionLevel: 45
```

**Expected behavior**
It seems like the first command should not default to zip since the image is bigger than 16 x 16 and the width and height are powers of 2

**Platform information:**
 - OIIO branch/version:  Installed oiiotool from brew
 - OS: MacOS

(I can open a PR, I just want to be sure there's no specific reason for this behavior)

**Comment by lgritz** [comment]
Interesting.

I am presuming that the switch to zip is intended to apply only to tiled files, since the logic only mentions tile_width and tile_height, and not the image size at all.

But note the clause

        && ((spec.tile_width < 16 && spec.tile_height < 16)

because for non-tiled files, the tile_width and tile_height are 0, so all single channel non-tiled channels are getting swept up in this.

If we believe that this is only a restriction for tiled files, then I would recommend changing that whole clause to

    // For single channel tiled images, dwaa/b compression only seems to work
    // reliably when tile size > 16 and size is a power of two.
    if (spec.nchannels == 1 && spec.tile_width != 0
        && Strutil::istarts_with(comp, "dwa")
        && ((spec.tile_width < 16 && spec.tile_height < 16)
            || !ispow2(spec.tile_width) || !ispow2(spec.tile_height))) {
        comp = "zip";
    }

which by checking `tile_width != 0` restricts the whole action to tiled images.

Now then, there are two important questions remaining:

1. Is there also a problem with non-tiled (scanline) single-channel images with dwa compression if the *image* size is < 16 or not a power of 2? We should check empirically, and have an example in the testsuite to prove it. If it is a problem (which we have been avoiding only due to inadvertently excluding *all* 1-channel scanline images from dwa compression!), then the logic needs to be adjusted to also make the compression switch for non-tiled images with the right properties, for example by adding this after tile case

  ```
    // For single channel scanline images, dwaa/b compression only seems to work
    // reliably when image size > 16 and size is a power of two.
    if (spec.nchannels == 1 && spec.width == 0
        && Strutil::istarts_with(comp, "dwa")
        && ((spec.width < 16 && spec.height < 16)
            || !ispow2(spec.width) || !ispow2(spec.height))) {
        comp = "zip";
    }
  ```

But this is something that needs to be checked, I really don't know if it's a problem. (@kdt3rd @peterhillman do you have any thoughts?)


2. Whatever the status of whether this is also a problem for non-tiled images, we might want to also establish whether it's true for all/current versions of OpenEXR. If it affects the current or fairly recent OpenEXR releases, then we need to keep all this logic. If it is only very old versions affected, we might debate at which point we should restrict this clause to only certain exr versions, or eliminate it altogether (for example, if it was fixed before the minimum openexr version that we claim to support).

**Comment by lgritz** [comment]
@RoseHirigoyen It would be very helpful if you could try to determine whether the suppression of dwa compression for 1-channel images should or should not be taken for scanline images (based on image size) as well as for tiled images (based on the tile size), and also confirm whether it's still an issue at all for the latest OpenEXR release, then make the appropriate PR with whatever logic changes seem to be called for by those results.

**Comment by lgritz** [comment]
Ah, this is a addressed by https://github.com/OpenImageIO/oiio/pull/3884

I knew that one felt like I had recently fixed the issue! I had not, but I had discussed it here, that's why it seemed familiar.

Closing because I do think #3884 should fix your problem. It's fixed in master and in the in-progress dev-2.4 branch which will be release 2.4.13.0 which I expect to be tagged and released on or very close to July 1.
